src/libsystemd/sd-journal/journal-file.c

```
@@ -543,6 +574,9 @@ static int journal_file_verify_header(JournalFile *f) {
 
                 if (f->header->field_hash_table_size == 0 || f->header->data_hash_table_size == 0)
                         return -EBADMSG;
+
+                if (JOURNAL_FILE_COMPRESSION(f) != compression_requested())
+                        return -EPROTONOSUPPORT;
```

If a file is using a given compression algorithm, and we support it, why not just continue using the file with that algorithm?

---

OK. But in that case, we need to explicitly rotate journal to enable the specified algorithm.

---

Yeah, I think that's totally fine. (Even after rotation, the old file with the old compression is going to be there. So setting the variable does not make everything use the new compression algorithm anyway.)

---

please return a recognizable error and then treat this as a reason for rotation further up the stack